### PR TITLE
switch typha to go modules from glide

### PR DESCRIPTION
**Pull request by rafaelvanoni**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -187,41 +185,30 @@ build-all: $(addprefix sub-build-,$(VALIDARCHES))
...
         echo "Old version: $$OLD_VER";\
 	if [ "$(LIBCALICO_VERSION)" != $$OLD_VER ]; then \
-            sed -i "s/$$OLD_VER/$(LIBCALICO_VERSION)/" glide.yaml && \
+            sed -i "s/$$OLD_VER/v0.0.0 $(LIBCALICO_VERSION)/" go.mod && \
 	    if [ "$(LIBCALICO_REPO)" != "github.com/projectcalico/libcalico-go" ]; then \
-              glide mirror set https://github.com/projectcalico/libcalico-go $(LIBCALICO_REPO) --vcs git; glide mirror list; \
+	      echo "replace github.com/projectcalico/libcalico-go => $(LIBCALICO_REPO) v0.0.0-$(LIBCALICO_VERSION)" >> go.mod; \
             fi;\
-          glide up --strip-vendor || glide up --strip-vendor; \
+          go mod vendor; \
```
Comment (spikecurtis) [inline Makefile]: This make target fails
```
$ make update-libcalico
mkdir -p .go-pkg-cache && docker run --rm --net=host -e GO111MODULE=on -v /run/user/1000/keyring/ssh:/ssh-agent --env SSH_AUTH_SOCK=/ssh-agent -e LOCAL_USER_ID=1000 -v /home/spike/repos/typha:/github.com/projectcalico/typha:rw -v /home/spike/repos/typha/.go-pkg-cache:/go-cache:rw -e GOCACHE=/go-cache -w /github.com/projectcalico/typha -e GOARCH=amd64 calico/go-build:v0.23 sh -c '\
        echo "Updating libcalico to  from github.com/projectcalico/libcalico-go"; \
        export OLD_VER=$(grep libcalico-go /github.com/projectcalico/typha/go.mod | awk '{print $2}') ;\
        echo "Old version: $OLD_VER";\
if [ "" != $OLD_VER ]; then \
            sed -i "s/$OLD_VER/v0.0.0 /" go.mod && \
    if [ "github.com/projectcalico/libcalico-go" != "github.com/projectcalico/libcalico-go" ]; then \
      echo "replace github.com/projectcalico/libcalico-go => github.com/projectcalico/libcalico-go v0.0.0-" >> go.mod; \
            fi;\
          go mod vendor; \
        fi'
Starting with UID : 1000
}) ;\
        echo "Old version: $OLD_VER";\
if [ "" != $OLD_VER ]; then \
            sed -i "s/$OLD_VER/v0.0.0 /" go.mod && \
    if [ "github.com/projectcalico/libcalico-go" != "github.com/projectcalico/libcalico-go" ]; then \
      echo "replace github.com/projectcalico/libcalico-go => github.com/projectcalico/libcalico-go v0.0.0-" >> go.mod; \
            fi;\
          go mod vendor; \
        fi: line 3: syntax error: unexpected end of file (expecting ")")
make: *** [Makefile:202: update-libcalico] Error 2
```

File: Makefile [hunk 2]
```diff
@@ -187,41 +185,30 @@ build-all: $(addprefix sub-build-,$(VALIDARCHES))
...
-	$(DOCKER_RUN) $(CALICO_BUILD) glide up --strip-vendor
-	touch vendor/.up-to-date
-
 # vendor is a shortcut for force rebuilding the go vendor directory.
 .PHONY: vendor
 vendor: vendor/.up-to-date
-vendor/.up-to-date: glide.lock
-	mkdir -p $$HOME/.glide
-	$(DOCKER_RUN) $(CALICO_BUILD) glide install --strip-vendor
+vendor/.up-to-date: go.mod
```
Comment (spikecurtis) [inline Makefile]: Does Typha still need the vendor directory?

Comment (rafaelvanoni) [reply Makefile]: Most (or all) of our repos rely on the vendor directory for local builds. As far as I can see that's the only reason we still need vendoring.

Comment (spikecurtis) [reply Makefile]: Noncontainer builds should use go modules as well, in which case vendor isn't needed.  For example, after a `make clean` I can still run `go build -o calico-typha cmd/calico-typha/typha.go` and it builds fine.

File: Makefile (hunk 1 above)
Comment (rafaelvanoni) [reply Makefile]: Fixed

File: Makefile [hunk 3]
```diff
@@ -187,41 +185,31 @@ build-all: $(addprefix sub-build-,$(VALIDARCHES))
...
-	if [ "$(LIBCALICO_VERSION)" != $$OLD_VER ]; then \
-            sed -i "s/$$OLD_VER/$(LIBCALICO_VERSION)/" glide.yaml && \
+        echo "Updating libcalico-go to $(LIBCALICO_VERSION) from $(LIBCALICO_REPO)"; \
+        echo "Old version: $(LIBCALICO_OLDVER)";\
+	if [ "$(LIBCALICO_VERSION)" != "$(LIBCALICO_OLDVER)" ]; then \
+	    sed -i "/libcalico-go/d" go.mod go.sum; \
+	    go get $(LIBCALICO_REPO)@$(LIBCALICO_VERSION); \
 	    if [ "$(LIBCALICO_REPO)" != "github.com/projectcalico/libcalico-go" ]; then \
-              glide mirror set https://github.com/projectcalico/libcalico-go $(LIBCALICO_REPO) --vcs git; glide mirror list; \
+	      echo "replace github.com/projectcalico/libcalico-go => $(LIBCALICO_REPO) v0.0.0-$(LIBCALICO_VERSION)" >> go.mod; \
```
Comment (spikecurtis) [inline Makefile]: I don't think you can hardcode `v0.0.0-` into the version.  I just ran the update to `68e175bba1ebf524e54a4d0034119e4cf146bffb` and got the resulting version:

```
github.com/projectcalico/libcalico-go v3.9.0-0.dev.0.20190801150727-68e175bba1eb+incompatible
```

File: Makefile (hunk 2 above)
Comment (rafaelvanoni) [reply Makefile]: I meant LOCAL_BUILDS, or the ones that use ../libcalico-go (or other local repositories).

File: Makefile (hunk 3 above)
Comment (rafaelvanoni) [reply Makefile]: I'm replacing that "echo" command with 
```go mod edit -replace github.com/projectcalico/libcalico-go=$(LIBCALICO_REPO)@$(LIBCALICO_VERSION)```
I think that should do it w/o all the v0.0.0 problems.

File: Makefile (hunk 2 above)
Comment (spikecurtis) [reply Makefile]: Use a `replace` command in `go.mod` to point to a local source library, rather than the `vendor` directory.

Comment (rafaelvanoni) [reply Makefile]: Ok, I think I got this working now. Also removed all the vendoring code. Please take a look (search for local_build in the Makefile).

**Pull request merged**